File: 2023/rust/day1/src/pt2.rs
use std::collections::HashMap;

// pub fn run(data: &String) {
//     let radix = 10;
//     let map = HashMap::from([
//         ("one", 1),
//         ("two", 2),
//         ("three", 3),
//         ("four", 4),
//         ("five", 5),
//         ("six", 6),
//         ("seven", 7),
//         ("eight", 8),
//         ("nine", 9),
//     ]);
//
//     // get nums in each line
//     let mut line_nums = Vec::new();
//     for line in data.lines() {
//         let mut nums = Vec::new();
//         let mut word = String::new();
//         // println!("{}", line);
//
//         // look l->r on the line for the first nu
//         for c in line.chars() {
//             word.push(c);
//             // check if one of the map keys is in the word
//             for k in map.keys() {
//                 if word.contains(k) {
//                     let n = *map.get(k).unwrap();
//                     nums.push(n);
//                     // println!("{}: {}", word, n);
//                     word.clear();
//                     break;
//                 }
//             }
//             if c.is_ascii_digit() {
//                 nums.push(c.to_digit(radix).unwrap());
//                 // println!("{}: {}", c, c);
//                 word.clear();
//             }
//         }
//         line_nums.push(nums);
//         // break;
//     }
//     // convert the nums in each line into numbers [1-99]
//     let mut sum = 0;
//     for nums in line_nums.iter() {
//         let tens = 10 * nums[0];
//         let ones = nums.last().unwrap();
//         sum += tens + ones;
//     }
//     println!("pt2: {}\n", sum);
// }

pub fn run(data: &String) {
    let radix = 10;
    let map = HashMap::from([
        ("one", 1),
        ("two", 2),
        ("three", 3),
        ("four", 4),
        ("five", 5),
        ("six", 6),
        ("seven", 7),
        ("eight", 8),
        ("nine", 9),
    ]);

    // get nums in each line
    let mut line_nums = Vec::new();
    for line in data.lines() {
        let mut nums = Vec::new();
        let mut word = String::new();
        // println!("{}", line);

        // look l->r on the line for the first nu
        for c in line.chars() {
            word.push(c);
            // check if one of the map keys is in the word
            for k in map.keys() {
                if word.contains(k) {
                    let n = *map.get(k).unwrap();
                    nums.push(n);
                    // println!("{}: {}", word, n);
                    word.clear();
                    break;
                }
            }
            if c.is_ascii_digit() {
                nums.push(c.to_digit(radix).unwrap());
                // println!("{}: {}", c, c);
                word.clear();
            }
            if nums.len() == 1 {
                word.clear();
                break;
            }
        }

        // look r->l on line for last num
        for c in line.chars().rev() {
            word = c.to_string() + &word;
            // check if one of the map keys is in the word
            for k in map.keys() {
                if word.contains(k) {
                    let n = *map.get(k).unwrap();
                    nums.push(n);
                    // println!("{}: {}", word, n);
                    word.clear();
                    break;
                }
            }
            if c.is_ascii_digit() {
                nums.push(c.to_digit(radix).unwrap());
                // println!("{}: {}", c, c);
                word.clear();
            }

            if nums.len() == 2 {
                break;
            }
        }
        line_nums.push(nums);
        // break;
    }
    // convert the nums in each line into numbers [1-99]
    let mut sum = 0;
    for nums in line_nums.iter() {
        let tens = 10 * nums[0];
        let ones = nums.last().unwrap();
        sum += tens + ones;
    }
    println!("pt2: {}\n", sum);
}

File: 2023/cpp/day1/pt2.cpp
#include <cstdio>
#include <fstream>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

using std::ifstream;
using std::printf;
using std::string;
using std::unordered_map;

const unordered_map<string, int> hashmap = {
    {"one", 1}, {"two", 2},   {"three", 3}, {"four", 4}, {"five", 5},
    {"six", 6}, {"seven", 7}, {"eight", 8}, {"nine", 9},
};
void pt2(ifstream& dataFile) {
    int sum = 0;
    int l = 0;
    string line;
    string word = "";

    while (std::getline(dataFile, line)) {
        std::cout << line << std::endl;
    }
    return;
    // for each line
    while (std::getline(dataFile, line)) {
        printf("line: %s\n", line.c_str());
        std::vector<int> nums;
        for (int i = 0; i < line.length(); i++) {
            word += line[i];
            if (std::isdigit(line[i])) {
                // std::cout << "c: " << line[i] << std::endl;
                nums.push_back(line[i] - '0');
                word.clear();
                // word = "";
            } else {
                for (const std::pair<const std::string, int>& n : hashmap) {
                    if (word.contains(n.first)) {
                        // printf("map %s, %s\n", word.c_str(),
                        // n.first.c_str());
                        nums.push_back(n.second);
                        // word = "";
                        word.clear();
                        break;
                    }
                }
            }
        }

        l++;
        for (int i = 0; i < nums.size(); i++) {
            printf("%d ", nums[i]);
        }
        sum += nums[0] * 10 + nums[nums.size() - 1];
        printf("%d\n", nums[0] * 10 + nums[nums.size() - 1]);
        std::cout << std::endl;
    }
    printf("pt2: %d - %s\n", sum, sum == 55093 ? "ok" : "not right");
}
